### PR TITLE
refactor type parser

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] others

## Modification
Refactor type parser:
1. Reduce type checking in `AstBuilder`
2. Reduce switch case in `Type.java`

### CODE REVIEW

**Comment by mofeiatwork** [comment]
run starrocks_be_unittest

**Comment by mofeiatwork** [comment]
run starrocks_clang-format

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/catalog/ScalarType.java [hunk 1]
```diff
@@ -152,61 +153,67 @@ public static ScalarType createType(PrimitiveType type) {
...
+            return ScalarType.createDecimalV3Type(PrimitiveType.DECIMAL128);
+        } else {
+            throw new IllegalArgumentException("Unsupported type specification: " + type.getText());
+        }
+    }
+
+    /**
+     * Create type from parser
+     */
+    public static Type createType(StarRocksParser.TypeContext type) {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ScalarType.java:175]: StarRocksParser.TypeContext itself is Antlr's internal data structure, which has nothing to do with the outside. This is a data structure that should be encapsulated inside AstBuilder. And now bringing it to ScalarType.java, there is nowhere to reuse this code, by spilling the types used inside Parser into other modules of the code, I don't think it's a good design

Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/ScalarType.java:175]: ok, we could reverse this dependency, let the parser depends on `Type.java`. Same as before, most code could be reused.

Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/catalog/ScalarType.java:175]: I totally agree

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 67 / 75 (89.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|28|34|82.35%|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|10|12|83.33%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Type.java|18|18|100.00%|
|:large_blue_circle:|com/starrocks/analysis/BinaryPredicate.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 2]
```diff
@@ -124,7 +129,7 @@
                     .addAll(floatPointTypes)
                     .build();
 
-    private static final ImmutableList<Type> supportedTypes =
+    protected static final ImmutableList<Type> supportedTypes =
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:132]:
```suggestion
    protected static final ImmutableList<Type> SUPPORT_TYPES =
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java [hunk 3]
```diff
@@ -112,6 +113,25 @@
                     .addAll(STRING_TYPE_LIST)
                     .build();
 
+    private static final ImmutableSortedSet<String> VARIABLE_TYPE_SET =
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java:116]: Why some use sort collector and some don't?

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 4]
```diff
@@ -147,6 +152,25 @@
...
+    protected static final ImmutableSortedMap<String, ScalarType> STATIC_TYPE_MAP =
+            ImmutableSortedMap.<String, ScalarType>orderedBy(String.CASE_INSENSITIVE_ORDER)
+                    .put("DECIMAL", Type.DEFAULT_DECIMALV2) // generic name for decimal
+                    .put("STRING", Type.DEFAULT_STRING)
+                    .put("INTEGER", Type.INT)
+                    .putAll(SUPPORT_SCALAR_TYPE_LIST.stream()
+                            .collect(Collectors.toMap(x -> x.getPrimitiveType().toString(), x -> (ScalarType) x)))
+                    .build();
+
+    protected static final ImmutableMap<PrimitiveType, ScalarType> PRIMITIVE_TYPE_SCALAR_TYPE_IMMUTABLE_MAP =
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:167]:
```suggestion
    protected static final ImmutableMap<PrimitiveType, ScalarType> PRIMITIVE_TYPE_SCALAR_TYPE_MAP =
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java (hunk 3 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java:116]: It's not for sorting, but for case-insensitive.

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java (hunk 2 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Type.java]: done

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java (hunk 4 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Type.java]: done

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 5]
```diff
@@ -147,6 +152,25 @@
...
                     .build();
 
+    protected static final ImmutableList<Type> SUPPORT_SCALAR_TYPE_LIST =
+            ImmutableList.copyOf(SUPPORTED_TYPES.stream().filter(Type::isScalarType).collect(Collectors.toList()));
+
+    protected static final ImmutableSortedMap<String, ScalarType> STATIC_TYPE_MAP =
+            ImmutableSortedMap.<String, ScalarType>orderedBy(String.CASE_INSENSITIVE_ORDER)
+                    .put("DECIMAL", Type.DEFAULT_DECIMALV2) // generic name for decimal
+                    .put("STRING", Type.DEFAULT_STRING)
+                    .put("INTEGER", Type.INT)
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:162]: Why add "INTEGER" alone, it looks like can not create Type.INT by this name before

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Review by HangyuanLiu — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java (hunk 5 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:162]: `INTEGER` is an alias of `INT`. There's some test cases cover it.

Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:162]: The parser could accept this token, in `AstBuilder.java:1548`.

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 68 / 76 (89.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|28|34|82.35%|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|10|12|83.33%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Type.java|19|19|100.00%|
|:large_blue_circle:|com/starrocks/analysis/BinaryPredicate.java|1|1|100.00%|

**Review by Seaven — APPROVED** [review]

**Pull request merged**